// @vssue/utils/lib/index.js
import format from 'date-fns/format';
import { parse, stringify } from 'qs';
export const buildURL = (cleanURL, params) => {
    const query = stringify(params, { addQueryPrefix: true });
    return `${cleanURL}${query}`;
};
export const concatURL = (baseURL, path) => {
    const firstPart = baseURL.replace(/\/$/, '');
    const secondPart = path.replace(/^\//, '');
    return `${firstPart}/${secondPart}`;
};
export const getCleanURL = (fullURL) => {
    const noHash = fullURL.split('#')[0] || '';
    const cleanURL = noHash.split('?')[0] || '';
    return cleanURL;
};
export const parseQuery = (URL) => parse(URL, { ignoreQueryPrefix: true });
export const buildQuery = (params) => stringify(params);
export const formatDateTime = (str) => {
    const dateTime = format(str, 'YYYY-MM-DD HH:mm:ss');
    return dateTime;
};
// eslint-disable-next-line @typescript-eslint/no-empty-function
export const noop = () => { };
export default {
    buildURL,
    concatURL,
    getCleanURL,
    parseQuery,
    buildQuery,
    formatDateTime,
    noop,
};
